### PR TITLE
feat(cast): Run Arbitrary Bytecode

### PR DESCRIPTION
**Pull request by abigger87**
## Overview

This PR introduces a new subcommand for cast: `run`.

## Motivation

Execute arbitrary bytecode using cast like the `hevm exec` command.

## Solution Checklist

- [x] Eat Bytecode
- [ ] Pass Bytecode into the EVM
- [ ] Return Pretty Results

### CODE REVIEW

**Comment by gakonst** [comment]
I like this directionally. It can be very powerful with forked mode.

One idea:
1. What if we put this in cast so that it's similar to `seth run-tx`, behind `cast run` or `cast exec`?
2. Should we add this functionality behind `cast send --simulate` as well, so that it runs a transaction against mainnet fork state?
3. Should we also allow passing a `<txhash>` which would simulate a forked transaction locally?

This would allow us to print traces when simulating transactions, which would be quite useful for people doing on-chain forensics and whatnot, given that ethtx.info does not exist for other chains, and people end up [using e.g. Tenderly's tracer/flamegraph](https://twitter.com/danielvf/status/1503756428212936710)

**Comment by abigger87** [comment]
> 

Open to making it a subcommand in `cast`, though I do believe @onbjerg makes a good argument that cast does not have any EVM context and this could bloat cast:
https://github.com/gakonst/foundry/issues/899#issuecomment-1065912974

**Comment by gakonst** [comment]
Ah yes, but imagine that someone doing operations against mainnet wouldn't be using forge, they'd be using cast right? So if the exec bytecode use case is in the context of doing ops, it makes sense to bundle it in cast? So I'd err towards putting it there, and reconsider if we see the binary bloating (altho couple megabytes added should be fine)?

File: cli/src/cmd/exec.rs [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+use ansi_term::Colour;
+use evm_adapters::evm_opts::{BackendKind, EvmOpts};
+use foundry_config::{figment::Figment, Config};
+
+// Loads project's figment and merges the build cli arguments into it
+foundry_config::impl_figment_convert!(ExecArgs, opts, evm_opts);
+
+#[derive(Debug, Clone, Parser)]
+pub struct ExecArgs {
+    #[clap(help = "the bytecode to execute")]
```
Comment (mattsse) [inline cli/src/cmd/exec.rs:19]: add note about expected format, hex

**Review by mattsse — COMMENTED** [review]
looking god so far.

on cast/forge, since both live in the same crate, we can use commands in both cast and forge, perhaps we can restructure the cmd mod into

`mod shared; mod forge; mod cast`

**Comment by tynes** [comment]
> 3\. Should we also allow passing a `<txhash>` which would simulate a forked transaction locally?
> 
> This would allow us to print traces when simulating transactions, which would be quite useful for people doing on-chain forensics and whatnot, given that ethtx.info does not exist for other chains, and people end up [using e.g. Tenderly's tracer/flamegraph](https://twitter.com/danielvf/status/1503756428212936710)

I'm a big fan of this idea, would prefer to use local + scriptable tooling over tenderly. I think one gotcha would be that if a tx hash is passed, then all previous txs in the block would need to be fetched and executed first in case a previous tx in the block altered the same state that the passed tx touches

**Comment by mds1** [comment]
> I'm a big fan of this idea, would prefer to use local + scriptable tooling over tenderly. I think one gotcha would be that if a tx hash is passed, then all previous txs in the block would need to be fetched and executed first in case a previous tx in the block altered the same state that the passed tx touches

Yea good call, this is how [`seth debug <txHash>`](https://github.com/dapphub/dapptools/blob/master/src/seth/libexec/seth/seth-debug) behaves, though in practice the execution of the previous txs with HEVM was too slow so I'd mainly just use `seth run-tx <txHash> --trace` or `seth run-tx <txHash> --debug` which skips the execution of prior txs. 

I think we should follow a similar pattern, and since there's a lot of overlap in the various seth/hevm commands, my proposal to clean up the API would be:
1. Keep `forge` as the test runner (i.e. no changes to forge here)
2. In this PR, setup `cast run` so it can execute arbitrary bytecode like [`hevm exec`](https://github.com/dapphub/dapptools/tree/master/src/hevm#hevm-exec)
3. In future PRs, update `cast run` to:
    1. Accept transaction hashes (instead of just bytecode)
    2. Support both `--trace` and `--debug` modes
    3. When passing a tx hash, execute previous txs in the block by default unless a flag is passed to skip execution of those txs. Potential names for that flag are  `--unsound`, `--unsafe`, `--rapid`, `--quick`
    4. For all touched addresses, pull from Etherscan by default for better traces/debugging, but also allow a `--source` or similar flag to pass the path to a forge directory containing contracts that should be compiled and referenced in the trace (e.g. for when a contract is not verified on Etherscan but you have it locally)

**Comment by abigger87** [comment]
Tackling as part of https://github.com/gakonst/foundry/pull/1020

**Pull request closed without merging**